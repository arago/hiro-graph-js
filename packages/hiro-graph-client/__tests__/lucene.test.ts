/* eslint-env jest */
/**
 *  Testing the Lucene Query Generator
 */
import {
  lucene as parse,
  getPlaceholderKeyForIndex,
  Lucene,
} from '../src/lucene';

describe('Lucene Query Generator:', function () {
  const tests = [
    {
      name: 'single field',
      input: { 'ogit/_id': 'xyz' },
    },
    {
      name: 'multiple field',
      input: {
        'ogit/_id': 'xyz',
        'ogit/_type': 'ogit/Comment',
      },
    },
    {
      name: 'multiple values (as and)',
      input: {
        field: ['value1', 'value2'],
      },
    },
    {
      name: '$not',
      input: { $not: { field: 'value' } },
    },
    {
      name: '$not (multi value)',
      input: { $not: { prop: ['test', 'not'] } },
    },
    {
      name: '$or (single value, trick question!)',
      input: { $or: { prop: 'test' } },
    },
    {
      name: '$or (multiple. same key)',
      input: { $or: { prop: ['test', 'value'] } },
    },
    {
      name: '$or (multiple. diff key)',
      input: { $or: { prop: 'test', anotherProp: 'this too' } },
    },
    {
      name: '$missing (single)',
      input: { $missing: 'prop' },
    },
    {
      name: '$missing (multiple)',
      input: { $missing: ['prop', 'anotherProp'] },
    },
    {
      name: '$range',
      input: { $range: { prop: [1, 5] as [number, number] } },
    },
    {
      name: '$search',
      input: { $search: `test "quoted term"` },
    },
    {
      name: '$or with $search',
      input: { $or: { '/target': 'test', $search: 'test2' } },
    },
    {
      name: 'wildcard',
      input: { '/a': 'test*', '/b': '*test', '/c': '*test*' },
    },
    {
      name: 'extreme example',
      input: {
        key1: 'value',
        key2: ['multi', 'value'],
        $not: {
          key4: 'not',
          key5: ['not', 'multi'],
          $range: {
            key6: ['notFrom', 'notTo'],
            key7: ['second', 'range'],
          },
          $or: { key8: ['not', 'or', 'values'] },
        },
        $or: {
          key9: 'or this',
          key10: ['two', 'terms'],
          $not: {
            key11: 'nor this',
          },
          $missing: 'key12',
          $search: 'or search',
        },
        $missing: ['key13', 'key14'],
        $search: 'test "quoted term"',
      },
    },
    {
      name: 'strings with quotes',
      input: { prop: 'test " quoted' },
    },
    {
      name: 'strings with existing slashes',
      input: { prop: 'test \\ slashed' },
    },
    {
      name: 'strings with exsiting slashed quotes',
      input: { prop: 'test \\" slashquoted' },
    },
    {
      name: 'long ngram',
      input: {
        $search:
          'this is a really long message, this wont split because of spaces',
      },
    },
    {
      name: 'ngram quoted',
      input: {
        $search: 'this is a quoted "message"',
      },
    },
    {
      name: 'ngram special',
      input: {
        $search:
          'ZTE_vEMS_1_U31_OMMOID=jorand1b-c#@#%#611-308111-1_1_1_1_198094420_327680_1602716595130-1603329369065',
      },
    },
    {
      name: 'long ngram split',
      input: {
        $search:
          'asdksdfidfhdffs423423dxisepfksdklgdlijpsdmgld456456fsdfsfesfdfsefsdfsefsdfsef1231435675ugljl',
      },
    },
    {
      name: 'long ngram split manual',
      input: {
        'ogit/_content.ngram':
          'asdksdfidfhdffs423423dxisepfksdklgdlijpsdmgld456456fsdfsfesfdfsefsdfsefsdfsef1231435675ugljl',
      },
    },
  ];

  tests.forEach(
    ({
      name,
      input,
      placeholders = [],
    }: {
      name: string;
      input: Lucene.Query;
      placeholders?: any[];
    }) => {
      it(name, function () {
        if (input) {
          const actual = parse(input);

          expect(actual.querystring).toMatchSnapshot();
          expect(actual.placeholders).toMatchSnapshot();
          placeholders.forEach((p, i) => {
            expect(actual.placeholders[getPlaceholderKeyForIndex(i)]).toEqual(
              p,
            );
          });
        }
      });
    },
  );
});
