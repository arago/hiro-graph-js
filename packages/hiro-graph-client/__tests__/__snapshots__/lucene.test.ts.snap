// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lucene Query Generator: $missing (multiple) 1`] = `"+_missing_:\\"prop\\" +_missing_:\\"anotherProp\\""`;

exports[`Lucene Query Generator: $missing (single) 1`] = `"+_missing_:\\"prop\\""`;

exports[`Lucene Query Generator: $not (multi value) 1`] = `"+(-prop:\\"test\\" -prop:\\"not\\")"`;

exports[`Lucene Query Generator: $not 1`] = `"+(-field:\\"value\\")"`;

exports[`Lucene Query Generator: $or (multiple. diff key) 1`] = `"+(prop:\\"test\\" anotherProp:\\"this too\\")"`;

exports[`Lucene Query Generator: $or (multiple. same key) 1`] = `"+(prop:\\"test\\" prop:\\"value\\")"`;

exports[`Lucene Query Generator: $or (single value, trick question!) 1`] = `"+(+prop:\\"test\\")"`;

exports[`Lucene Query Generator: $range 1`] = `"+prop:[\\"1\\" TO \\"5\\"]"`;

exports[`Lucene Query Generator: $search 1`] = `"+_content.ngram:$ph_0"`;

exports[`Lucene Query Generator: extreme example 1`] = `"+key1:\\"value\\" +(key2:\\"multi\\" key2:\\"value\\") +(-key4:\\"not\\" -key5:\\"not\\" -key5:\\"multi\\" -key6:[\\"notFrom\\" TO \\"notTo\\"] -key7:[\\"second\\" TO \\"range\\"] -(key8:\\"not\\" key8:\\"or\\" key8:\\"values\\")) +(key9:\\"or this\\" key10:\\"two\\" key10:\\"terms\\" (-key11:\\"nor this\\") _missing_:\\"key12\\" _content.ngram:$ph_0) +_missing_:\\"key13\\" +_missing_:\\"key14\\" +_content.ngram:$ph_1"`;

exports[`Lucene Query Generator: multiple field 1`] = `"+ogit\\\\/_id:\\"xyz\\" +ogit\\\\/_type:\\"ogit/Comment\\""`;

exports[`Lucene Query Generator: multiple values (as and) 1`] = `"+(field:\\"value1\\" field:\\"value2\\")"`;

exports[`Lucene Query Generator: single field 1`] = `"+ogit\\\\/_id:\\"xyz\\""`;

exports[`Lucene Query Generator: strings with existing slashes 1`] = `"+prop:\\"test \\\\\\\\ slashed\\""`;

exports[`Lucene Query Generator: strings with exsiting slashed quotes 1`] = `"+prop:\\"test \\\\\\\\\\\\\\" slashquoted\\""`;

exports[`Lucene Query Generator: strings with quotes 1`] = `"+prop:\\"test \\\\\\" quoted\\""`;
